substrate-node/pallets/pallet-smart-contract/src/storage_state/v9.rs

```diff
@@ -0,0 +1,978 @@
+    // Nothing to do here
+}
+
+fn clean_rent_contract<T: Config>() {
+    // Nothing to do here
+}
+
+// ContractsToBillAt
+pub fn clean_contracts_to_bill_at<T: Config>() -> frame_support::weights::Weight {
+    debug!(
+        "🧼  Smart Contract pallet {:?} cleaning ContractsToBillAt storage map START",
+        PalletVersion::<T>::get()
+    );
+
+    let mut r = 0u64;
+    let mut w = 0u64;
+
+    let contracts_to_bill_at = ContractsToBillAt::<T>::iter().collect::<Vec<_>>();
+    r.saturating_accrue(contracts_to_bill_at.len() as u64);
+    let mut contract_id_count = vec![0; (ContractID::<T>::get() + 1) as usize];
+    r.saturating_inc();
+
+    for (_index, contract_ids) in contracts_to_bill_at.clone() {
+        for contract_id in contract_ids {
+            contract_id_count[contract_id as usize].saturating_inc();
+        }
+    }
+
+    for (index, contract_ids) in contracts_to_bill_at {
+        let mut new_contract_ids = Vec::new();
+        for contract_id in contract_ids {
+            if let Some(contract) = Contracts::<T>::get(contract_id) {
+                match contract.contract_type {
+                    types::ContractData::NodeContract(node_contract) => {
+                        let contract_resource = NodeContractResources::<T>::get(contract_id);
+                        r.saturating_inc();
+                        if node_contract.public_ips > 0 || !contract_resource.used.is_empty() {
+                            if contract_id_count[contract_id as usize] == 1 {
+                                new_contract_ids.push(contract.contract_id);
+                            }
+                            contract_id_count[contract_id as usize].saturating_dec();
+                        }
+                    }
+                    _ => {
+                        if contract_id_count[contract_id as usize] == 1 {
+                            new_contract_ids.push(contract.contract_id);
+                        }
+                        contract_id_count[contract_id as usize].saturating_dec();
+                    }
+                };
+            }
+            r.saturating_inc();
+        }
+        ContractsToBillAt::<T>::insert(index, new_contract_ids);
+        w.saturating_inc();
+    }
+
+    debug!(
+        "✨  Smart Contract pallet {:?} cleaning ContractsToBillAt storage map END",
+        PalletVersion::<T>::get()
+    );
+
+    T::DbWeight::get().reads_writes(r.saturating_add(2), w)
+}
+
+// ActiveNodeContracts
+pub fn clean_active_node_contracts<T: Config>() -> frame_support::weights::Weight {
+    debug!(
+        "🧼  Smart Contract pallet {:?} cleaning ActiveNodeContracts storage map START",
+        PalletVersion::<T>::get()
+    );
+
+    let mut r = 0u64;
+    let mut w = 0u64;
+
+    let active_node_contracts = ActiveNodeContracts::<T>::iter().collect::<Vec<_>>();
+    r.saturating_accrue(active_node_contracts.len() as u64);
+
+    for (node_id, mut contract_ids) in active_node_contracts {
```

Comment:
Translate would work here no?

Reply:
I think it would work but not sure if it is the best choice.
https://paritytech.github.io/substrate/master/frame_support/storage/types/struct.StorageMap.html#method.translate
Even if it removes element when returning `None`, since we don't change the type and just change the value, maybe it does some unnecessary operation (i.e transforming the storage format)
See also https://docs.substrate.io/reference/how-to-guides/storage-migrations/basic-storage-migration/#write-migrate_to_v2

Reply:
I think we can `mutate` the storage map in place, the goal here is to remove objects and mutate the storage map.

Reply:
what we want here is change value + element removal

`mutate` => change value (not enough)
`translate` => change type + element removal (too much)

so I don t see the real benefit of using one of them unless I miss something...
or maybe doing 2 loops (1) removal (2) mutate

Reply:
inside the mutate you can call remove on the storage map
```
map.mutate(|x| {
   if x == toRemove {
       map.remove(x)
   }
})
```

Reply:
and this will result in one loop

Reply:
Tested and for StorageMap the mutate is done by key so it doesn't do "internal" loop:
https://paritytech.github.io/substrate/master/frame_support/storage/types/struct.StorageMap.html#method.mutate
We still have to do the main loop from outside so I don t see the benefit